webhook/pr_review.go

```diff
@@ -0,0 +1,384 @@
+			time.Sleep(time.Duration(sleepTimeSecond) * time.Second)
+			builder.Reset()
+		}
+	}
+}
+
+func getChatGPTComment(patch string) (string, error) {
+	client, err := openai.New(
+		openai.WithToken(viper.GetString("openai.api_key")),
+		openai.WithModel(viper.GetString("openai.model")),
+		openai.WithOrgID(viper.GetString("openai.org_id")),
+		openai.WithProxyURL(viper.GetString("openai.proxy")),
+		openai.WithSocksURL(viper.GetString("openai.socks")),
+		openai.WithBaseURL(viper.GetString("openai.base_url")),
+		openai.WithTimeout(viper.GetDuration("openai.timeout")),
+		openai.WithMaxTokens(viper.GetInt("openai.max_tokens")),
+		openai.WithTemperature(float32(viper.GetFloat64("openai.temperature"))),
+	)
+	if err != nil {
+		log.Println("Error creating OpenAI client:", err)
+		return "", err
+	}
+
+	out, err := util.GetTemplateByString(
+		prompt.CodeReviewTemplate,
+		util.Data{
+			"file_diffs": patch,
+		},
+	)
+	if err != nil {
+		log.Println("Error creating prompt:", err)
+		return "", err
+	}
+
+	ctx := context.Background()
+	resp, err := client.Completion(ctx, out)
+	if err != nil {
+		log.Println("Error completing prompt:", err)
+		return "", err
+	}
+	summarizeMessage := resp.Content
+
+	log.Println("PromptTokens: " + strconv.Itoa(resp.Usage.PromptTokens) +
+		", CompletionTokens: " + strconv.Itoa(resp.Usage.CompletionTokens) +
+		", TotalTokens: " + strconv.Itoa(resp.Usage.TotalTokens))
+
+	// Output core review summary
+	log.Println("================Review Summary====================")
+	log.Println("\n" + strings.TrimSpace(summarizeMessage) + "\n\n")
+	log.Println("==================================================")
+
+	return summarizeMessage, nil
+}
+
+func getLastLinePosition(patch string) int {
+	lines := strings.Split(patch, "\n")
+	position := 0
+	for _, line := range lines {
+		if strings.HasPrefix(line, "@@") {
+			continue
+		}
+		position++
+	}
+	return position
+}
+
+func submitReviewComment(owner, repo string, prNumber int, commitID, filePath, comment string, position int) error {
+	client := &http.Client{}
+
+	commentRequest := CommentRequest{
+		Body:     comment,
+		Path:     filePath,
+		Position: position,
+		CommitID: commitID,
+	}
+
+	jsonData, err := json.Marshal(commentRequest)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest("POST", fmt.Sprintf(submitCommentsURL, apiBaseurl, owner, repo, prNumber), bytes.NewBuffer(jsonData))
+	if err != nil {
+		log.Println("Error creating request:", err)
+		return err
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Authorization", "token "+token)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Println("Error submitting review comment:", err)
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusCreated {
+		log.Println("Review comment submitted successfully")
+	} else {
+		return fmt.Errorf("error: %d", resp.StatusCode)
+	}
+
+	return nil
+}
```

Comment:
Overall, the code patch seems relatively safe and simple, with no obvious security vulnerabilities or major bugs. However, there are a few potential areas for improvement:

- The code does not handle any potential errors or exceptions that may occur when making the HTTP request, besides logging and returning the error. It may be useful to add more detailed error handling and reporting, such as retrying failed requests or providing more specific error messages.
- The code assumes that the authorization token is already obtained and passed as a parameter to the function. However, it is unclear how this token is obtained or stored, and whether it may be exposed or compromised in some way.
- The code only sets two HTTP headers (Content-Type and Authorization), but does not specify any other parameters or options for the request. Depending on the API endpoint and use case, there may be other headers or parameters that should be included or customized.
- The code does not explicitly check or handle any potential security risks or attacks, such as cross-site scripting (XSS) or injection attacks. Depending on the context and requirements of the API, it may be necessary to add more security measures or input validation.